envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt

```diff
@@ -22,6 +22,15 @@ class SnapshotUpdater(
         private val logger by logger()
     }
 
+    @Volatile
+    private var groups: List<Group> = listOf()
+    @Volatile
+    private var trigger: String = ""
```

Comment:
I think it would be better to not use mutable volatile state, but incorporate `groups` and `trigger` in the object that is passed in Flux itself.

Reply:
UPDATE: I see it would be hard to detect which source produced event in combineLatest

Maybe we should do it another way:
Don't combine `onGroupAdded ` and `changes` - process them separately. The logic for each of these events is different:
* on `changes` event - generate global snapshots and update snapshots for all groups
* on `onGroupAdded` - just update snapshot for changed groups using `lastSnapshot`